### PR TITLE
Fix CLI help formatting

### PR DESCRIPTION
**Pull request by newhoggy**

### CODE REVIEW

**Comment by newhoggy** [comment]
Before:

```
Usage: cardano-cli transaction build-raw [--byron-era | --shelley-era |
                                           --allegra-era | --mary-era |
                                           --alonzo-era] (--tx-in TX-IN
                                         [--tx-in-script-file FILE
                                           [
                                             (--tx-in-datum-file FILE |
                                               --tx-in-datum-value JSON VALUE)
                                             (--tx-in-redeemer-file FILE |
                                               --tx-in-redeemer-value JSON VALUE)
                                             --tx-in-execution-units (INT, INT)]])
                                         [--tx-in-collateral TX-IN]
                                         [--tx-out ADDRESS VALUE
                                           [--tx-out-datum-hash HASH]]
                                         [--mint VALUE (--mint-script-file FILE
                                           [
                                             (--mint-redeemer-file FILE |
                                               --mint-redeemer-value JSON VALUE)
                                             --mint-execution-units (INT, INT)])]
                                         [--invalid-before SLOT]
                                         [--invalid-hereafter SLOT]
                                         [--fee LOVELACE]
                                         [--certificate-file CERTIFICATEFILE
                                           [--certificate-script-file FILE
                                             [
                                               (--certificate-redeemer-file FILE |

                                                 --certificate-redeemer-value JSON VALUE)
                                               --certificate-execution-units (INT, INT)]]]
                                         [--withdrawal WITHDRAWAL
                                           [--withdrawal-script-file FILE
                                             [
                                               (--withdrawal-redeemer-file FILE |

                                                 --withdrawal-redeemer-value JSON VALUE)
                                               --withdrawal-execution-units (INT, INT)]]]
                                         [--json-metadata-no-schema |
                                           --json-metadata-detailed-schema]
                                         [--auxiliary-script-file FILE]
                                         [--metadata-json-file FILE |
                                           --metadata-cbor-file FILE]
                                         [--genesis FILE |
                                           --protocol-params-file FILE]
                                         [--update-proposal-file FILE]
                                         --out-file FILE
```

After:

```
Usage: cardano-cli transaction build-raw  
            [ --byron-era
            | --shelley-era
            | --allegra-era
            | --mary-era
            | --alonzo-era
            ]
            ( --tx-in TX-IN
              [ --tx-in-script-file FILE
                [ (--tx-in-datum-file FILE | --tx-in-datum-value JSON VALUE)
                  ( --tx-in-redeemer-file FILE
                  | --tx-in-redeemer-value JSON VALUE
                  )
                  --tx-in-execution-units (INT, INT)
                ]
              ]
            )
            [--tx-in-collateral TX-IN]
            [--tx-out ADDRESS VALUE [--tx-out-datum-hash HASH]]
            [ --mint VALUE
              ( --mint-script-file FILE
                [ (--mint-redeemer-file FILE | --mint-redeemer-value JSON VALUE)
                  --mint-execution-units (INT, INT)
                ]
              )
            ]
            [--invalid-before SLOT]
            [--invalid-hereafter SLOT]
            [--fee LOVELACE]
            [ --certificate-file CERTIFICATEFILE
              [ --certificate-script-file FILE
                [ ( --certificate-redeemer-file FILE
                  | --certificate-redeemer-value JSON VALUE
                  )
                  --certificate-execution-units (INT, INT)
                ]
              ]
            ]
            [ --withdrawal WITHDRAWAL
              [ --withdrawal-script-file FILE
                [ ( --withdrawal-redeemer-file FILE
                  | --withdrawal-redeemer-value JSON VALUE
                  )
                  --withdrawal-execution-units (INT, INT)
                ]
              ]
            ]
            [--json-metadata-no-schema | --json-metadata-detailed-schema]
            [--auxiliary-script-file FILE]
            [--metadata-json-file FILE | --metadata-cbor-file FILE]
            [--genesis FILE | --protocol-params-file FILE]
            [--update-proposal-file FILE]
            --out-file FILE
```

**Comment by newhoggy** [comment]
In order to get the traced version of the output for inspection in a browser, supply `CLI_HELP_TRACE=1` in the environment.  For example:

```
CLI_HELP_TRACE=1 cardano-cli transaction build-raw
```

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 1]
```diff
@@ -557,7 +558,11 @@ pTransaction :: Parser TransactionCmd
...
   asum
     [ subParser "build-raw"
-        (Opt.info pTransactionBuildRaw $ Opt.progDesc "Build a transaction (low-level, inconvenient)")
+        $ Opt.info pTransactionBuildRaw $ Opt.progDescDoc $ Just $
+          pretty @String "Build a transaction (low-level, inconvenient)"
+          <> line <> line
+          <> "Please note the order of some cmd options is crucial. If used incorrectly may produce "
+          <> "undesired tx body. See nested [] notation above for details."
     , subParser "build"
         (Opt.info pTransactionBuild $ Opt.progDesc "Build a balanced transaction (automatically calculates fees)")
```
Comment (james-iohk) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:567]: We also want the new text from `build-raw` for this new `build` cmd.

**Comment by james-iohk** [comment]
@newhoggy Much improved overall. There's just a couple of things bugging me and causing misalignment. Neither of these are serious.

1. Only sometimes is there's a space after `(` or `[`, before the option name. I see both `( --tx-in-redeemer-file FILE` (with space) and `(--tx-in-datum-file FILE` (without space). This occurs throughout for all cmds. I think it looks cleaner with the space.
2. The nested options for `--tx-in-script-file` are not aligned. The opening bracket for `--tx-in-redeemer-file` and `--tx-in-redeemer-value` etc are 1 space ahead of the opening bracket for `--tx-in-datum-file`. See:
```
            (--tx-in TX-IN
              [--tx-in-script-file FILE
                [(--tx-in-datum-file FILE | --tx-in-datum-value JSON VALUE)
                  ( --tx-in-redeemer-file FILE
                  | --tx-in-redeemer-value JSON VALUE
                  )
                  --tx-in-execution-units (INT, INT)]])
```

**Comment by newhoggy** [comment]
Yeah, ideally, I want it to look like this:

```
           ( --tx-in TX-IN
              [ --tx-in-script-file FILE
                [ (--tx-in-datum-file FILE | --tx-in-datum-value JSON VALUE)
                  ( --tx-in-redeemer-file FILE
                  | --tx-in-redeemer-value JSON VALUE
                  )
                  --tx-in-execution-units (INT, INT)]])
```

ie. space after the bracket/parens if the group spans more than one line, but no space otherwise.  Unfortunately the layout code is fairly complex and I haven't yet worked it out even after spending over a day on it.  The whole layout code may need a rewrite to achieve the right effect.

I'm interest in merging this on the basis that it is an improvement.

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 2]
```diff
@@ -557,9 +558,17 @@ pTransaction :: Parser TransactionCmd
...
-        (Opt.info pTransactionBuildRaw $ Opt.progDesc "Build a transaction (low-level, inconvenient)")
+        $ Opt.info pTransactionBuildRaw $ Opt.progDescDoc $ Just $
+          pretty @String "Build a transaction (low-level, inconvenient)"
+          <> line <> line
+          <> "Please note the order of some cmd options is crucial. If used incorrectly may produce "
+          <> "undesired tx body. See nested [] notation above for details."
     , subParser "build"
-        (Opt.info pTransactionBuild $ Opt.progDesc "Build a balanced transaction (automatically calculates fees)")
+        $ Opt.info pTransactionBuild $ Opt.progDescDoc $ Just $
+          pretty @String "Build a transaction (low-level, inconvenient)"
```
Comment (james-iohk) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:568]: This should be `"Build a balanced transaction (automatically calculates fees)"`

File: moo.html [hunk 3]
```diff
@@ -0,0 +1 @@
+zsh: command not found: transaction
```
Comment (james-iohk) [inline moo.html:1]: We want this file?

**Review by james-iohk — CHANGES_REQUESTED** [review]

**Review by james-iohk — APPROVED** [review]

File: cabal.project [hunk 4]
```diff
@@ -107,6 +107,18 @@ package cardano-ledger-alonzo-test
 -- --------------------------- 8< --------------------------
 -- Please do not put any `source-repository-package` clause above this line.
 
+source-repository-package
```
Comment (Jimbo4350) [inline cabal.project:110]: Can we leave a comment that we want to eventually not rely on our own fork after the relevant PR has been merged?

File: cardano-cli/src/Cardano/CLI/Render.hs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import           Prelude (String)
+import           Prettyprinter
+import           Prettyprinter.Render.Util.SimpleDocTree
+
+import qualified Data.Text as T
+import qualified System.Environment as IO
+import qualified System.IO.Unsafe as IO
+
+cliHelpTraceEnabled :: Bool
+cliHelpTraceEnabled = IO.unsafePerformIO $ do
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Render.hs:19]: I don't have any experience using `unsafePerformIO`. Is this ok to use here?

**Review by Jimbo4350 — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/Render.hs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  mValue <- IO.lookupEnv "CLI_HELP_TRACE"
+  return $ mValue == Just "1"
+{-# NOINLINE cliHelpTraceEnabled #-}
```
Comment (erikd) [inline cardano-cli/src/Cardano/CLI/Render.hs:22]: Yes we need that.

File: cardano-cli/src/Cardano/CLI/Render.hs (hunk 5 above)
Comment (erikd) [reply cardano-cli/src/Cardano/CLI/Render.hs:19]: Yes this seems to be a safe use of `unsafePerformIO` .

Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Render.hs:19]: Yep.  This example of `unsafePerformIO` is fairly normal.  `unsafePerformIO` is used to perform some IO to initialise a global constant.  The danger is if the side effect is executed more than once, in which case the constant is not so constant, but this is prevent by the `NOINLINE` pragma.  This forces the value to be a thunk that can only be evaluated once.

Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Render.hs:19]: See documentation on `unsafePerformIO`: https://hackage.haskell.org/package/base-4.15.0.0/docs/System-IO-Unsafe.html

**Comment by newhoggy** [comment]
bors merge

**Review by MarcFontaine — APPROVED** [review]
Thanks for fixing the tests.
Looks good to me

**Comment by newhoggy** [comment]
bors merge

**Pull request merged**